component/loki/source/api/internal/lokipush/push_api_server.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+	promql_parser "github.com/prometheus/prometheus/promql/parser"
+)
+
+type PushAPIServer struct {
+	logger       log.Logger
+	serverConfig *fnet.ServerConfig
+	server       *fnet.TargetServer
+	handler      loki.EntryHandler
+
+	rwMutex       sync.RWMutex
+	labels        model.LabelSet
+	relabelRules  []*relabel.Config
+	keepTimestamp bool
+}
+
+func NewPushAPIServer(logger log.Logger,
+	serverConfig *fnet.ServerConfig,
+	handler loki.EntryHandler,
+	registerer prometheus.Registerer,
+) (*PushAPIServer, error) {
+	s := &PushAPIServer{
+		logger:       logger,
+		serverConfig: serverConfig,
+		handler:      handler,
+	}
+
+	srv, err := fnet.NewTargetServer(logger, "agent_loki_source_api", registerer, serverConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	s.server = srv
+	return s, nil
+}
+
+func (s *PushAPIServer) Run() error {
+	level.Info(s.logger).Log("msg", "starting push API server")
+
+	err := s.server.MountAndRun(func(router *mux.Router) {
+		router.Path("/api/v1/push").Methods("POST").Handler(http.HandlerFunc(s.handleLoki))
+		router.Path("/api/v1/raw").Methods("POST").Handler(http.HandlerFunc(s.handlePlaintext))
+		router.Path("/ready").Methods("GET").Handler(http.HandlerFunc(s.ready))
+	})
+	return err
+}
+
+func (s *PushAPIServer) ServerConfig() fnet.ServerConfig {
+	return *s.serverConfig
+}
+
+func (s *PushAPIServer) Shutdown() {
+	level.Info(s.logger).Log("msg", "stopping push API server")
+	s.server.StopAndShutdown()
+}
+
+func (s *PushAPIServer) SetLabels(labels model.LabelSet) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.labels = labels
+}
+
+func (s *PushAPIServer) getLabels() model.LabelSet {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.labels.Clone()
+}
+
+func (s *PushAPIServer) SetKeepTimestamp(keepTimestamp bool) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.keepTimestamp = keepTimestamp
+}
+
+func (s *PushAPIServer) getKeepTimestamp() bool {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.keepTimestamp
+}
+
+func (s *PushAPIServer) SetRelabelRules(rules frelabel.Rules) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.relabelRules = frelabel.ComponentToPromRelabelConfigs(rules)
+}
+
+func (s *PushAPIServer) getRelabelRules() []*relabel.Config {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	newRules := make([]*relabel.Config, len(s.relabelRules))
+	for i, r := range s.relabelRules {
+		var rCopy = *r
+		newRules[i] = &rCopy
+	}
+	return newRules
+}
+
+// NOTE: This code is copied from Promtail (3478e180211c17bfe2f3f3305f668d5520f40481) with changes kept to the minimum.
+// Only the HTTP handler functions are copied to allow for flow-specific server configuration and lifecycle management.
+func (s *PushAPIServer) handleLoki(w http.ResponseWriter, r *http.Request) {
```

Comment:
I am in favour of adopting this approach instead - we can simplify a lot of surrounding code and only copy the handler functions.